### ISSUE TITLE
How can I use `AsyncClient *client = new AsyncClient();` with a simple `AsyncWebServer server(80);` ?

### ISSUE THREAD

**Issue by GeorgeFlorian** [me-no-dev/ESPAsyncWebServer#553]
Board: ESP32

## TL;DR

> I would like to gain access to `client.onTimeout([](void*, AsyncClient*, uint32_t time) {} );` but I don't know how to use `AsyncClient *client = new AsyncClient();`.
> 
> I would really like to know how to use `client.onTimeout([](void*, AsyncClient*, uint32_t time) {});` , because, even though I can live with using `ESP.restart()` for each `ack timeout 4` since they are rare, the number of `rx timeout 4` warning is fairly bigger and very few of them actually freeze the server, so I can't restart the ESP that frequently because I won't be able to do a thing with it.

## Long story:

Hello !

I don't have a WebSocket in my project, nor do I understand it's purpose/use.
I would have to remake the whole project in order to add WebSocket.
I'm currently using handlers for all GETs and POSTs for each SPIFFS file inside the ESP.

For example:
```
      server.on("/dhcpIP", HTTP_GET, [](AsyncWebServerRequest *request){
        if(userFlag) {
          if(!request->authenticate(value_login[0].c_str(), value_login[1].c_str()))
            return request->requestAuthentication(NULL,false);            
          request->send(SPIFFS, "/dhcpIP.html", "text/html");
        } else {
          request->send(SPIFFS, "/dhcpIP.html", "text/html");
        }
      });
      server.on("/dhcpIP", HTTP_POST, [](AsyncWebServerRequest * request){        
        int params = request->params();
        String values_dhcp[2];
        for(int i=0;i<params;i++){
          AsyncWebParameter* p = request->getParam(i);
          if(p->isPost()){
              logOutput((String)"POST[" + p->name().c_str() + "]: " + p->value().c_str() + "\n");            
              values_dhcp[i] = p->value();            
            } else {
                logOutput((String)"GET[" + p->name().c_str() + "]: " + p->value().c_str() + "\n");
              }
        } // for(int i=0;i<params;i++)
        
        if(values_dhcp[0] != NULL && values_dhcp[0].length() != 0 &&
          values_dhcp[1] != NULL && values_dhcp[1].length() != 0) {
          File inputsWrite = SPIFFS.open("/networkRelay.txt", "w");
          if(!inputsWrite) logOutput((String)"ERROR_INSIDE_POST ! Couldn't open file to write DHCP IP credentials !");
          inputsWrite.println(values_dhcp[0]);  // SSID
          inputsWrite.println(values_dhcp[1]);  // Password
          inputsWrite.close();
          logOutput("Configuration saved !");
          request->send(200, "text/html", "<div style=\"text-align:center; font-family:arial;\">Congratulation !</br></br>You have successfully changed the networks settings.</br></br>The device will now restart and try to apply the new settings.</br></br>You can get this device's IP by looking through your AP's DHCP List."); 
          shouldReboot = true;
        } else request->redirect("/dhcpIP");
      
      }); // server.on("/dhcpLogin", HTTP_POST, [](AsyncWebServerRequest * request)
```

I also have no idea if WebSocket is what I need, because, apart from all the web-pages on the server, I also need simple GET methods to be able to control the ESP's relays from another piece of software using only their URLs: 
```
      server.on("/relay1/on", HTTP_GET, [](AsyncWebServerRequest *request){
        if(userFlag) {
          if(!request->authenticate(value_login[0].c_str(), value_login[1].c_str()))
            return request->requestAuthentication(NULL,false);    
          digitalWrite(RELAY1, HIGH);
          status1 = "ON";
          logOutput((String)relay1 + " is ON");
          if(Delay1.toInt() == 0) {
            needManualCloseRelayOne = true;
            logOutput((String)relay1 + " will remain open until it is manually closed !");
          } else {
            needManualCloseRelayOne = false;
            logOutput((String)relay1 + " will automatically close in " + Delay1 + " seconds !");
          }          
          startTimeRelayOne = millis();
          // Serial.println("Do I get here ? /relay1/on");
          request->send(200, "text/plain", relay1 + " is ON");
        } else {    
            digitalWrite(RELAY1, HIGH);
            status1 = "ON";
            logOutput((String)relay1 + " is ON");
            if(Delay1.toInt() == 0) {
              needManualCloseRelayOne = true;
              logOutput((String)relay1 + " will remain open until it is manually closed !");
            } else {
              needManualCloseRelayOne = false;
              logOutput((String)relay1 + " will automatically close in " + Delay1 + " seconds !");
            }
            startTimeRelayOne = millis();
            // Serial.println("Do I get here ? /relay1/on");
            request->send(200, "text/plain", relay1 + " is ON");
        }
      });

      server.on("/relay1/off", HTTP_GET, [](AsyncWebServerRequest *request){
        if(userFlag) {
          if(!request->authenticate(value_login[0].c_str(), value_login[1].c_str()))
            return request->requestAuthentication(NULL,false);    
          digitalWrite(RELAY1, LOW);
          status1 = "OFF";
          logOutput((String)relay1 + " is OFF");
          needManualCloseRelayOne = true;
          // Serial.println("Do I get here ? /relay1/off");
          request->send(200, "text/plain", relay1 + " is OFF");
        } else {    
            digitalWrite(RELAY1, LOW);
            status1 = "OFF";
            logOutput((String)relay1 + " is OFF");
            needManualCloseRelayOne = true;
            // Serial.println("Do I get here ? /relay1/off");
            request->send(200, "text/plain", relay1 + " is OFF");
        }
      });
```

I've tried adding `AsyncClient *client = new AsyncClient();` outside of `void setup() {}` but when using `Serial.println(aClient->connected());` or `Serial.println(aClient->localIP().toString());` I always get 0 or 0.0.0.0. It's like I don't have any client.

This is surely because I don't attach the `client` to nothing.

How can I use this ? There isn't any example.

**Comment by philbowles** [comment]
??? the AsyncClient *client = new AsyncClient(); needs to be in a _different sketch_ on a _different machine_ . It needs to be in the code that connects to this server, not in this server itself.

AsyncClient is used by code that requests a TCP connection to a server, i.e. it is a client (hence the name). I looked at your related issue. It appears to me you have a fundamental misunderstanding of the client-server model and/or what code runs where...

We are talking about two totally separate MCUs here

CLIENT -------------------------------------------->SERVER
creates AsyncClient.......................................................runs AsyncWebserver
connects to: ------------------------------------>
<------------------------------------------------- replies to client
asks for data (usually web page)--------------->
<------------------------------------------------ sends web page or 404 etc

Both of these objects  happen to use the underlying asyncTCP library (each with its own copy in its own mcu), one to send, one to receive, don't mix up the different functionality just because they share a common library

Otherwise, what am I missing? One of us is very confused.

**Comment by AkhileshThorat** [comment]
Hi @philbowles 
I think the problem @GeorgeFlorian encounters is the way to connect to a remote webSocket server(different ESP32). As he said there aren't any examples demonstrating how to connect a Client(first ESP32 ) to a server(second esp32 running async web server) and vice versa.
Even I'm unable to figure out the API's for connecting to a remote server.
Lets say my server is running on http://{ESP-IP}/ws, what should I write to establish a connection with this endpoint?

**Comment by jhoughjr** [comment]
I also have found the proper way to make a client connection, and I really need that functionality. I'll keep working on it and if I figure it out, I'll post it here.

**Comment by GeorgeFlorian** [comment]
> I also have found the proper way to make a client connection, and I really need that functionality. I'll keep working on it and if I figure it out, I'll post it here.

Great
